_inc/lib/class.jetpack-service-helper.php

```diff
@@ -0,0 +1,162 @@
+
+		if ( $service ) {
+			$url = add_query_arg( array( 'service' => $service ), $url );
+		}
+
+		if ( count( $params ) ) {
+			$url = add_query_arg( $params, $url );
+		}
+
+		return $url;
+	}
+
+	static function connect_url( $service_name ) {
+		return add_query_arg( array(
+			'action'   => 'request',
+			'service'  => $service_name,
+			'kr_nonce' => wp_create_nonce( 'keyring-request' ),
+			'nonce'    => wp_create_nonce( "keyring-request-$service_name" ),
+		), menu_page_url( 'sharing', false ) );
+	}
+
+	static function refresh_url( $service_name ) {
+		return add_query_arg( array(
+			'action'   => 'request',
+			'service'  => $service_name,
+			'kr_nonce' => wp_create_nonce( 'keyring-request' ),
+			'refresh'  => 1,
+			'for'      => 'publicize',
+			'nonce'    => wp_create_nonce( "keyring-request-$service_name" ),
+		), admin_url( 'options-general.php?page=sharing' ) );
+	}
+
+	static function disconnect_url( $service_name, $id ) {
+		return add_query_arg( array(
+			'action'   => 'delete',
+			'service'  => $service_name,
+			'id'       => $id,
+			'kr_nonce' => wp_create_nonce( 'keyring-request' ),
+			'nonce'    => wp_create_nonce( "keyring-request-$service_name" ),
+		), menu_page_url( 'sharing', false ) );
+	}
+
+	static function admin_page_load() {
+		if ( isset( $_GET['action'] ) ) {
+			if ( isset( $_GET['service'] ) ) {
+				$service_name = $_GET['service'];
+			}
+
+			switch ( $_GET['action'] ) {
+
+				case 'request':
+					check_admin_referer( 'keyring-request', 'kr_nonce' );
+					check_admin_referer( "keyring-request-$service_name", 'nonce' );
+
+					$verification = Jetpack::generate_secrets( 'publicize' );
+					if ( ! $verification ) {
+						$url = Jetpack::admin_url( 'jetpack#/settings' );
+						wp_die( sprintf( __( "Jetpack is not connected. Please connect Jetpack by visiting <a href='%s'>Settings</a>.", 'jetpack' ), $url ) );
+
+					}
+					$stats_options = get_option( 'stats_options' );
+					$wpcom_blog_id = Jetpack_Options::get_option( 'id' );
+					$wpcom_blog_id = ! empty( $wpcom_blog_id ) ? $wpcom_blog_id : $stats_options['blog_id'];
+
+					$user     = wp_get_current_user();
+					$redirect = Jetpack_Service_Helper::api_url( $service_name, urlencode_deep( array(
+						'action'       => 'request',
+						'redirect_uri' => add_query_arg( array( 'action' => 'done' ), menu_page_url( 'sharing', false ) ),
+						'for'          => 'publicize',
+						// required flag that says this connection is intended for publicize
+						'siteurl'      => site_url(),
+						'state'        => $user->ID,
+						'blog_id'      => $wpcom_blog_id,
+						'secret_1'     => $verification['secret_1'],
+						'secret_2'     => $verification['secret_2'],
+						'eol'          => $verification['exp'],
+					) ) );
+					wp_redirect( $redirect );
+					exit;
+					break;
+
+				case 'completed':
+					Jetpack::load_xml_rpc_client();
+					$xml = new Jetpack_IXR_Client();
+					$xml->query( 'jetpack.fetchPublicizeConnections' );
```

Comment:
Since this is no longer publicize-specific, should we create a new endpoint (which is really just an alias of the current endpoint) called `jetpack.fetchKeyringConnections`? Otherwise could be confusing in the future.

Similarly with `publicize_connections`, maybe it should be called `keyring_connections`, and on plugin upgrade we migrate the existing option name to the new option name.